control-plane/api-gateway/controllers/gatewayclass_controller.go

```diff
@@ -0,0 +1,101 @@
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	GatewayClassFinalizer      = "gateway-exists-finalizer.networking.x-k8s.io"
+	GatewayClassControllerName = "hashicorp.com/consul-api-gateway-controller"
+)
+
+// GatewayClassReconciler reconciles a GatewayClass object.
+// The GatewayClass is responsible for defining the behavior of API gateways
+// which reference the given class.
+type GatewayClassReconciler struct {
+	ControllerName string
+	Log            logr.Logger
+	client.Client
+}
+
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/finalizers,verbs=update
+
+// Reconcile handles the reconciliation loop for GatewayClass objects.
+func (r *GatewayClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("gatewayClass", req.NamespacedName)
+
+	gc := &gwv1beta1.GatewayClass{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, gc)
+	if err != nil {
+		log.Error(err, "unable to get GatewayClass")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if string(gc.Spec.ControllerName) != r.ControllerName {
+		// This GatewayClass is not for this controller.
```

Comment:
In case the controller name changed, we should clean up any finalizers we've added here so that the object can actually be deleted. I know that the particular finalizer name is part of the spec, but this is another place where the upstream spec seems a bit odd.

Particularly, if a gateway class switches names from a valid controller name to an invalid controller name and it was previously in use (but now no longer is because it's invalid and can't be reconciled by a non-existent controller), then the class can never be deleted.

I'm thinking we can probably just write our own finalizer and then clean it up if the name switched just to have a clean separation between controller implementations and to make sure we're cleaning up the stuff we've added to the object -- though I think this is probably worth raising with upstream and I'm sure @mikemorris probably has some thoughts on it.

Reply:
So the logic here is that we would define our own finalizer -- say, `consul-gateway-finalizer` and if the controller name does not match, go remove the finalizer?

Reply:
Yep, that's my thought, I know the spec is prescriptive about the finalizer name, but frankly, it doesn't really make sense to me due to the reasons above.

Reply:
I think a replacement. I'd probably prefer the replacement in this implementation, and given that Mike is currently out, if we have some strong push-back in terms of a conformance to upstream, then we can go back and switch it.